cli/src/alluxio.org/cli/processes/common.go

```diff
@@ -12,113 +12,150 @@
+		}
+		nodes = append(nodes, masterList...)
+	}
+	if onWorkers {
+		if len(workerList) == 0 {
+			if err := getNodes(false); err != nil {
+				log.Logger.Fatalf("Cannot get workers, error: %s", err)
+			}
+		}
+		nodes = append(nodes, workerList...)
+	}
+
+	// get public key if nones
+	if privateKeySigner == nil {
+		if err := getPrivateKey(); err != nil {
+			log.Logger.Fatalf("Cannot get private key, error: %s", err)
+		}
 	}
 
-	mastersReader := bufio.NewReader(mastersFile)
-	var mastersList []string
-	lastLine := false
-	for !lastLine {
-		// read lines of the workers file
-		line, err := mastersReader.ReadString('\n')
+	// dial nodes, run sessions and close
+	var errors []error
+	for _, node := range nodes {
+		conn, err := dialConnection(node, privateKeySigner)
 		if err != nil {
-			if err == io.EOF {
-				lastLine = true
-			} else {
-				log.Logger.Errorf("Error parsing worker file at this line: %s", line)
-				return nil, err
-			}
+			errors = append(errors, err)
+			continue
 		}
-		// remove notes
-		if strings.Index(line, "#") != -1 {
-			line = line[:strings.Index(line, "#")]
+		if err := runSession(node, conn, command); err != nil {
+			errors = append(errors, err)
 		}
-		line = strings.TrimSpace(line)
-		if line != "" {
-			mastersList = append(mastersList, line)
+		if err := conn.Close(); err != nil {
+			log.Logger.Infof("Connection to %s closed. Error: %s", node, err)
+		} else {
+			log.Logger.Infof("Connection to %s closed.", node)
 		}
 	}
-	return mastersList, nil
+
+	if len(errors) != 0 {
+		log.Logger.Warningf("Run command %s failed, number of failures: %v", command, len(errors))
+		return stacktrace.Propagate(errors[0], "First error: ")
+	}
+	log.Logger.Infof("Run command %s successful on nodes: %s", command, nodes)
+	return nil
 }
 
-func getWorkers() ([]string, error) {
-	workersDir := path.Join(env.Env.EnvVar.GetString(env.ConfAlluxioConfDir.EnvVar), "workers")
-	workersFile, err := os.Open(workersDir)
-	if err != nil {
-		log.Logger.Errorf("Error reading worker hostnames at %s", workersDir)
-		return nil, err
+func addStartFlags(argument string, cmd *env.StartProcessCommand) string {
+	if cmd.AsyncStart {
+		argument = argument + " -a"
 	}
+	if cmd.SkipKillOnStart {
+		argument = argument + " -N"
+	}
+	return cliPath + " " + argument
+}
 
-	workersReader := bufio.NewReader(workersFile)
-	var workersList []string
-	lastLine := false
-	for !lastLine {
-		// read lines of the workers file
-		line, err := workersReader.ReadString('\n')
-		if err != nil {
-			if err == io.EOF {
-				lastLine = true
-			} else {
-				log.Logger.Errorf("Error parsing worker file at this line: %s", line)
-				return nil, err
-			}
-		}
-		// remove notes
-		if strings.Index(line, "#") != -1 {
-			line = line[:strings.Index(line, "#")]
+func addStopFlags(argument string, cmd *env.StopProcessCommand) string {
+	if cmd.SoftKill {
+		argument = argument + " -s"
+	}
+	return cliPath + " " + argument
+}
+
+func getNodes(isMasters bool) error {
```

Comment:
using a boolean only works because there are 2 files. better to not assume this and pass in the file name as the argument